src/main/java/uk/gov/hmcts/cft/idam/testingsupportapi/service/NotificationsService.java

```diff
@@ -0,0 +1,119 @@
+package uk.gov.hmcts.cft.idam.testingsupportapi.service;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.cft.idam.api.v2.common.error.SpringWebClientHelper;
+import uk.gov.service.notify.Notification;
+import uk.gov.service.notify.NotificationClient;
+import uk.gov.service.notify.NotificationClientException;
+import uk.gov.service.notify.NotificationList;
+
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@Slf4j
+@Service
+public class NotificationsService {
+
+    private static final Pattern NOTIFY_KEY_REFERENCE_REGEX = Pattern.compile("^[a-z]*[_a-z]*[_a-z]*");
+    private static final Pattern OLDER_THAN_REGEX = Pattern.compile("older_than=([A-Za-z0-9\\-]+)&?");
+
+    private static final String ALL_STATUSES = null;
+    private static final String EMAIL_TYPE = "email";
+    private static final int MAX_PAGES = 20;
+
+    @Value("${notify.key}")
+    private String notifyKey;
+
+    private String notifyReference;
+
+    private final NotificationClient notificationClient;
+
+    public NotificationsService(NotificationClient notificationClient) {
+        this.notificationClient = notificationClient;
+    }
+
+    @VisibleForTesting
+    protected void setNotifyKey(String key) {
+        this.notifyKey = key;
+    }
+
+    /**
+     * Find latest notification.
+     * @should return latest notification
+     * @should throw http status code exception
+     */
+    public Optional<Notification> findLatestNotification(String email) throws Exception {
+        try {
+            return findEmailInNotifications(email);
+        } catch (NotificationClientException e) {
+            log.error("Exception while finding notification for '{}'", email, e);
+            throw SpringWebClientHelper.exception(HttpStatus.INTERNAL_SERVER_ERROR, e);
+        }
+    }
+
+    /**
+     * Find email in notifications.
+     * @should find first notification for email
+     * @should return empty if the email is not found and there are no more pages to search
+     * @should return empty if the email is not found and the page limit is reached
+     */
+    protected Optional<Notification> findEmailInNotifications(String searchEmail) throws NotificationClientException {
+        String olderThanNotificationId = null;
+        int page = 0;
+        while (page < MAX_PAGES) {
+            NotificationList currentPage = notificationClient
+                .getNotifications(ALL_STATUSES, EMAIL_TYPE, getNotifyReference(), olderThanNotificationId);
```

Comment:
[minor] `getNotifyReference()` could be a `private` property in the class as it will never change (no need to re-evaluate it on every page load).

Reply:
Yeah, I considered that, but I'm not entirely sure about the order of @Value versus a constructor, and I didn't want it in the constructor. Longer term I think we might want to change "reference" to be "<env>/<email>" since I think that means we could search for exactly the email we want.

Reply:
I think we can inject notifyKey through the constructor as well.

Reply:
I'm not keen on it, but let's see what it looks like

Reply:
It makes the mocking really awkward to initialise in the tests